### DISCUSSION TITLE
v0.10.0-beta.0

### DISCUSSION THREAD

**Discussion by sayanarijit** [sayanarijit/xplr#161] [General]
Changes & fixes
----------------------
- Added support for native lua binding via `~/.config/xplr/init.lua` ([see this](https://github.com/sayanarijit/xplr/pull/156)).
- Fixed open in editor action by only opening the file under focus.

<hr /><em>This discussion was created from the release <a href='https://github.com/sayanarijit/xplr/releases/tag/v0.10.0-beta.0'>v0.10.0-beta.0</a>.</em>

**Comment by sayanarijit** [comment]
This release is for people living on the edge to play around with lua native support by moving config from [`config.yml`](https://github.com/sayanarijit/xplr/blob/main/src/config.yml) to `init.lua`.

Syntax:

```lua
version = 'v0.10.0-beta.0'

config = xplr.config

config.general.default_ui.prefix = ' '
config.general.default_ui.suffix = ''

config.general.focus_ui.prefix = '▸'
config.general.focus_ui.suffix = ''
config.general.focus_ui.style.fg = { Rgb = { 170, 150, 130 } }
config.general.focus_ui.style.bg = { Rgb = { 50, 50, 50 } }
config.general.focus_ui.style.add_modifiers = { 'Bold' }

-- Remap
config.modes.builtin.default.key_bindings.on_key.e = config.modes.builtin.action.key_bindings.on_key.e

-- Remove (one of the lua surprises - setting key to `nil` removes the table entry)
config.modes.builtin.action.key_bindings.on_key.e = nil
```

There are further plans like adding plugin support, moving from handlebars to native lua, add `LuaExec` similar to [`BashExec`](https://docs.rs/xplr/0.9.1/xplr/app/enum.ExternalMsg.html#variant.BashExec) etc. but let's do them one-at-a-time, experimenting with different options. Also, If anyone managed to close [config.yml -> init.lua](https://github.com/sayanarijit/xplr/issues/160), it'd be awesome.

**Comment by sayanarijit** [comment]
I think that's it. It was quite a journey, but now I can say `xplr` has the basic foundation of Lua integration which allows us to overcome the yaml inheritance limitations/complexity and explore further customization options. Hence, once reviewed, `v0.10.0` can be released out in the open.

Notable changes since `v0.10.0-beta.0`:

- `config.yml` has been fully replaced with [`init.lua`](https://github.com/sayanarijit/xplr/blob/main/src/init.lua). If you have a lot of customization in your `config.yml`, [xplr-yml2lua](https://github.com/sayanarijit/xplr-yml2lua) can help you with migrating it to lua.
- `Handlebars` templates has been replaced with Lua functions. [e.g.](https://github.com/sayanarijit/xplr/blob/bfdb7736b99bc3c5ae53e7d621ba0e7ca2299b14/src/init.lua#L291-L308)
- `xplr` will fall back to default config if `init.lua` is corrupt. You will see the error via log. So, now, it's not a big deal if you mess up `init.lua` or you upgrade xplr and there's version incompatibility issue. You can still use it.
- Added (optional) mouse scrolling support, which be auto enabled on devices that support mouse.
- History navigation will now behave like jump list in vim. So, you even can jump and return to different locations in the same directory.
- Added messages `CallLua` and `CallLuaSilently` to call lua functions. The app state will be passed as inputs to the functions, and the returned messages will be handled by xplr. `CallLua` and `CallLuaSilently` are more flexible (and probably faster) alternatives to `Call`, `CallSilently`, `BashExec` and `BashExecSilently`. [e.g.](https://github.com/sayanarijit/xplr/pull/177#issue-650643573)

cc @maximbaz In case you missed this.

**Reply by max-baz** [reply]
Impressive work! I haven't had a chance to try it out, but definitely looking forward to! 😉

**Comment by sayanarijit** [comment]
Join https://github.com/sayanarijit/xplr/discussions/183 for further discussion.
